app.rb

```diff
@@ -16,15 +59,280 @@ class MyApp < Sinatra::Base
+    session.delete :logged_in_user_id
+    session.delete :token_handler
+    session.delete :state
+  end
+
+  def change_password newpassword
+    c = current_user
+    c.password = Digest::SHA256.hexdigest(newpassword)
+    c.save!
+  end
+
+  def token_string
+    session[:token_handler].access_token_type +
+      " " +
+      session[:token_handler].access_token
+  end
+
+  def current_user
+    if id = session[:logged_in_user_id]
+      User.find_by id: id
+    else
+      nil
+    end
+  end
+
+  def parsed_body
+    begin
+      @parsed_body ||= JSON.parse request.body.read
+    rescue
+      halt 400
+    end
+  end
+
+  get "/callback" do
+    if Digest::SHA256.hexdigest(session[:state]).to_s == params["state"]
+      session[:token_handler].request_refresh_and_access_tokens params
+      redirect "/dashboard"
+    else
+      halt 401
+    end
+  end
+
+  get "/api/me" do
+    session[:token_handler] = SpotifyApiToken.new
+    session[:state] = SecureRandom.random_number(2**1024).to_s
+    redirect(session[:token_handler].authorize! Digest::SHA256.hexdigest( session[:state] ))
+  end
+
+  post "/songs" do
+    @song = Song.new(title: params[:title], artist: params[:artist], suggester_id: current_user.id)
+    spotify = SpotifyApiRequest.new(
+      authorization: token_string,
+      song: @song.title
+    )
+    spotify.parse!
+    @hits = spotify.get_songs
+    if @hits.count == 0
+      session[:message] = "No Song Found"
+      redirect "/dashboard"
+    else
+      erb :choose_song
+    end
+    # rescue
+    #   status 403
+    #   # redirect "/dashboard"
+    #   halt "Entry doesn't include Title, Artist"
+  end
+
+  post "/choose_song" do
+    @song = Song.new(title: params[:title], artist: params[:artist], suggester_id: current_user.id, spotify_id: params[:spotify_id])
+    if (Playlist.current.map {|s| s.spotify_id}).include?(@song.spotify_id)
+      session[:message] = "Song Already On List"
+      redirect '/dashboard'
+    elsif @song.save!
+      200
+      session[:message] = "Song Added Successfully"
+      redirect '/dashboard'
+    else
+      status 403
+      session[:message] = "Song Could Not Be Added"
+      redirect '/dashboard'
+    end
+  end
+
+  post "/no_song" do
+  session[:message] = "No Song Was Added"
+  redirect '/dashboard'
+  end
+
+  delete "/songs" do
+    song = Song.where(title: params[:title], suggester_id: current_user.id)
+    song.delete_all
+    redirect "/dashboard"
+  end
+
+
+  get "/winningplaylist" do
+    @last_playlist = Playlist.last
+    playlist = Playlist.last
+    if playlist
+        @songs = playlist.songs
+        @last_5_playlists = Playlist.last(5)
+    end
+    time = playlist.created_at
+    twelve_hour_clock_time = '%m-%d-%Y'
+    @us_time = time.strftime(twelve_hour_clock_time)
+    erb :winningplaylist
+  end
+
+
+  get "/winningplaylist/:playlist_id" do
+    @playlist = Playlist.find_by(id: params[:playlist_id])
+    @songs = @playlist.songs
```

Comment:
If `find_by` fails, `@playlist` will be `nil` and `@playlist.songs` will error. You probably want to do `Playlist.find(params[:playlist_id])` instead - that way, if nothing is found, you'll raise an `ActiveRecord::RecordNotFound` and send back a 404 page.